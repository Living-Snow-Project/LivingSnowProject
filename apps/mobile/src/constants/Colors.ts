const tintColor = "#2f95dc";

export const Colors = {
  tintColor,
  tabIconDefault: "#0077ff",
  tabIconSelected: tintColor,
  tabBar: "#fefefe",
  errorBackground: "red",
  errorText: "#fff",
  warningBackground: "#EAEB5E",
  warningText: "#666804",
  noticeBackground: tintColor,
  noticeText: "#fff",
};
