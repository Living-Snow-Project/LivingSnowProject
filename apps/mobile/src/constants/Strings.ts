// for TextInput placeholder prop
export const Placeholders = {
  GPS: {
    AcquiringLocation: "Looking for GPS signal",
    EnterCoordinates: "ie. 12.34567, -123.45678",
    NoPermissions:
      "Check location permissions in Settings. Enter coordinates manually.",
    NoLocation: "Could not determine location. Enter coordinates manually.",
  },
  Settings: {
    Username: "Enter your name",
    Organization: "Enter organization you belong to (if any)",
  },
  RecordScreen: {
    LocationDescription: "ie: Blue Lake, North Cascades, WA",
    Notes: "ie. dark red algae on glacial ice",
    Size: "Select a size",
    TubeId: "Leave blank if tube does not have an id",
    TubeIdDisabled: "Disabled - only used for Samples",
  },
};

export const Labels = {
  Date: "Date",
  Delete: "Delete",
  RecordForm: {
    Color: "Color (select all that apply)",
    Gps: "GPS Coordinates (latitude, longitude)",
    Photos: "Select Photos (limit 4)",
    Type: "Are you Reporting a Sighting or Taking a Sample?",
  },
  // TODO: rename to RecordDetailsFields
  RecordFields: {
    Type: "Type",
    Name: "Name",
    Organization: "Organization",
    TubeId: "Tube Id",
    GPSCoordinates: "Coordinates",
    Latitude: "Latitude",
    Longitude: "Longitude",
    Size: "Size",
    Colors: "Colors",
    LocationDescription: "Location Description",
    Notes: "Notes",
    Photos: "Photos",
  },
  TimelineScreen: {
    ExampleRecords: "Example Record",
    PendingRecords: "Pending",
    DownloadedRecords: "Downloaded",
  },
  SettingsScreen: {
    DarkMode: "Dark",
    ManualCoordinates: "Manual Coordinates Warning",
  },
  StatusBar: {
    NoConnection: "No Internet Connection",
  },
  Modal: {
    Cancel: "Cancel",
    Confirm: "Confirm",
    DiskUsage: {
      header: "Confirm Delete",
      body: "Are you sure you want to delete saved photos?",
    },
    GpsManualEntry: {
      header: "Enter coordinates manually?",
      body: `Some users enter coordinates manually if they cannot acquire GPS signal or when they return home.\n\nThis message can be disabled in Settings.`,
    },
  },
};

// for Modal alerts
export const Notifications = {
  uploadSuccess: {
    title: "Record Uploaded",
    message: "Thanks for your submission.",
  },
  uploadRecordFailed: {
    title: "Record Saved",
    message: "We will upload it later.",
  },
  uploadPhotoFailed: {
    title: "Photo Saved",
    message: "We will upload it later.",
  },
  uploadPhotosFailed: {
    title: "Photos Saved",
    message: "We will upload them later.",
  },
  updateRecordSuccess: {
    title: "Record Updated",
  },
  updateRecordFailed: {
    title: "Update Failed",
    message:
      "This can happen when Internet connectivity is restored while updating the record.",
  },
  backgroundTasksNotAllowed: {
    title: "Background Uploads Not Enabled",
    message:
      "Please enable Background App Refresh if you would like stored records to be uploaded in the background.",
  },
};

export const Validations = {
  invalidCoordinates:
    'Coordinates must be in "lat, long" format. eg. 12.34567, -123.45678',
  invalidAlgaeSize: "A size must be specified",
  invalidAlgaeColors: "At least 1 color must be specified",
};

export const Errors = {
  recordNotFound: "Record to update not found",
};

export const RecordDescription = {
  Sample: `Sample`,
  Sighting: `Sighting`,
  Undefined: "Undefined",
};

export const AlgaeSizeDescription = {
  Select: "Select a size",
  Fist: "Fist",
  ShoeBox: "Shoe Box",
  CoffeeTable: "Coffee Table",
  Car: "Car",
  Bus: "Bus",
  Playground: "Playground",
  SportsField: "Sports Field",
  Other: "Other (describe in notes)",
};

export const AlgaeColorDescription = {
  Select: "Select a color",
  Red: "Red",
  Pink: "Pink",
  Grey: "Grey",
  Green: "Green",
  Orange: "Orange",
  Yellow: "Yellow",
  Other: "Other (describe in notes)",
};

export const BackgroundTasks = {
  UploadData: "background-upload-task",
};

// TODO: move this to Labels
export const Headers = {
  DiskUsage: "Disk Usage",
  Profile: "Profile",
  Prompts: "Prompts",
  Theme: "Theme",
};
