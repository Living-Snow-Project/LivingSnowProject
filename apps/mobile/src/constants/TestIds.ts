export const TestIds = {
  GPS: {
    GpsManualEntryModal: "gps-enable-manual-coordinates-modal",
    GpsManualEntryPressable: "gps-manual-pressable",
  },
  Selectors: {
    AlgaeColor: "algae-color-selector",
    AlgaeSize: "algae-size-selector",
    Photos: "photo-selector",
    RecordType: "record-type-selector",
  },
  RecordScreen: {
    UploadButton: "record-screen-upload-button",
    UpdateButton: "record-screen-update-button",
  },
  TimelineScreen: {
    FlatList: "timeline-screen-flatlist",
    RecordListView: "timeline-screen-record-list-view",
    NewRecordButton: "timeline-screen-record-button",
    SettingsButton: "timeline-screen-settings-button",
    ScrollToTopButton: "timeline-screen-scroll-to-top-button",
  },
  SettingsScreen: {
    ShowGpsWarning: "settings-screen-show-gps-warning",
  },
  Icons: {
    AddPhotosIcon: "add-photos-icon",
    DeleteIcon: "delete-icon",
    DoneSelectingPhotosIcon: "done-selecting-photos-icon",
    EditIcon: "edit-icon",
    PictureIcon: "picture-icon",
    SampleIcon: "sample-icon",
    SaveIcon: "save-icon",
    ScrollTopIcon: "scroll-top-icon",
    SightingIcon: "sighting-icon",
    SnowIcon: "snow-icon",
  },
  RecordList: {
    MenuTrigger: "pending-record-menu-trigger",
    EditRecordAction: "edit-record-action",
    DeleteRecordAction: "delete-record-action",
  },
  Modal: {
    ConfirmButton: "modal-confirm-button",
    DeletePendingRecord: "delete-pending-record",
    DiskUsageDeletePhotos: "disk-usage-delete-photos-modal",
    NoButton: "modal-no-button",
  },
};
