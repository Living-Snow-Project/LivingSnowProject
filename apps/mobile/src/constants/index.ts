export * from "./Colors";
export * from "./Strings";
export * from "./TestIds";
