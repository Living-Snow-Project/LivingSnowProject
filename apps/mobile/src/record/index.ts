export * from "./Color";
export * from "./Record";
export * from "./Size";
