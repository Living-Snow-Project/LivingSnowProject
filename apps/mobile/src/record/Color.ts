import { AlgaeColor } from "@livingsnow/record";
import { AlgaeColorDescription } from "../constants/Strings";

type AlgaeColorSelectorItem = {
  value: AlgaeColor;
  label: string;
  color?: { light: string; dark: string };
};

const algaeColorSelectorItems: AlgaeColorSelectorItem[] = [
  {
    value: "Red",
    label: AlgaeColorDescription.Red,
    color: { light: "red.700", dark: "red.500" },
  },
  {
    value: "Pink",
    label: AlgaeColorDescription.Pink,
    color: { light: "pink.500", dark: "pink.400" },
  },
  {
    value: "Green",
    label: AlgaeColorDescription.Green,
    color: { light: "green.700", dark: "green.400" },
  },
  {
    value: "Orange",
    label: AlgaeColorDescription.Orange,
    color: { light: "orange.500", dark: "orange.400" },
  },
  {
    value: "Yellow",
    label: AlgaeColorDescription.Yellow,
    color: { light: "yellow.600", dark: "yellow.500" },
  },
  {
    value: "Grey",
    label: AlgaeColorDescription.Grey,
    color: { light: "gray.600", dark: "gray.300" },
  },
  { value: "Other", label: AlgaeColorDescription.Other },
];

const getAllAlgaeColorSelectorItems = (): AlgaeColorSelectorItem[] =>
  algaeColorSelectorItems;

export { getAllAlgaeColorSelectorItems };
