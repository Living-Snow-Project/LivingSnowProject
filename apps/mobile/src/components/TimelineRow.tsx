import React from "react";
import { useNavigation } from "@react-navigation/native";
import { Box, HStack, Pressable, Text, VStack } from "native-base";
import { AlgaeRecord, recordDateFormat } from "@livingsnow/record";
import { RootStackNavigationProp } from "../navigation/Routes";
import { getUserStyle } from "./UserStyle";
import { PhotosLayout } from "./PhotosLayout";
import { Divider } from "./Divider";

function bottomText({
  locationDescription,
  notes,
}: AlgaeRecord): JSX.Element[] | null {
  const result: JSX.Element[] = [];
  if (locationDescription) {
    result.push(
      <Text key={0} fontWeight="500">
        {locationDescription}
      </Text>
    );
  }

  if (notes) {
    result.push(<Text key={1}>{notes}</Text>);
  }

  return result.length > 0 ? result : null;
}

function getRecordInfo(record: AlgaeRecord) {
  // TODO: useColorModeValue
  return (
    <HStack>
      <Text fontWeight={700} color="blue.700">
        {`${record.type}`}
      </Text>
      <Text fontWeight={400} color="blue.700">
        {` on ${recordDateFormat(record.date)}`}
      </Text>
    </HStack>
  );
}

type TimelineRowProps = {
  record: AlgaeRecord;
};

export function TimelineRow({ record }: TimelineRowProps) {
  const { navigate } = useNavigation<RootStackNavigationProp>();
  const { org, name, avatar } = getUserStyle(record.name, record.organization);

  return (
    <>
      <Pressable
        testID={record.id.toString()}
        onPress={() => navigate("RecordDetails", { record })}
      >
        <Box px={2} py={1} _dark={{ bg: "dark.100" }}>
          <VStack>
            <HStack>
              <Box width="15%">{avatar}</Box>
              <Box width="75%">
                <VStack ml={2}>
                  {name}
                  {org}
                  {getRecordInfo(record)}
                </VStack>
              </Box>
            </HStack>
            {bottomText(record)}
          </VStack>
        </Box>
        <PhotosLayout photos={record.photos} />
      </Pressable>
      <Divider />
    </>
  );
}
