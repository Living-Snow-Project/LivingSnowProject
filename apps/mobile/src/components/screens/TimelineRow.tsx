import React from "react";
import { useNavigation } from "@react-navigation/native";
import { Box, HStack, Pressable, Text, VStack } from "native-base";
import { AlgaeRecord, Photo } from "@livingsnow/record";
import { RootStackNavigationProp } from "../../navigation/Routes";
import { Divider, ThemedBox } from "../layout";
import { UserStyle } from "./UserStyle";
import { PhotosLayout } from "../media";
import { MinimalAlgaeRecord } from "../../../types";

function bottomText({
  locationDescription,
  notes,
}: AlgaeRecord): JSX.Element[] | null {
  const result: JSX.Element[] = [];
  if (locationDescription) {
    result.push(
      <Text key={0} fontWeight="500">
        {locationDescription}
      </Text>
    );
  }

  if (notes) {
    result.push(<Text key={1}>{notes}</Text>);
  }

  return result.length > 0 ? result : null;
}

type TimelineRowProps = {
  record: AlgaeRecord;
  photos?: Photo[];
  actionsMenu?: JSX.Element;
};

export function TimelineRow({ record, photos, actionsMenu }: TimelineRowProps) {
  const { navigate } = useNavigation<RootStackNavigationProp>();
  const recordDetail: MinimalAlgaeRecord = {
    record: { ...record },
    photos: photos ? [...photos] : undefined,
  };

  return (
    <>
      <Pressable
        testID={record.id}
        onPress={() =>
          navigate("RecordDetails", { record: JSON.stringify(recordDetail) })
        }
      >
        <ThemedBox px={2} py={1}>
          <VStack>
            <HStack>
              <UserStyle record={record} />
              <Box width="7%">{actionsMenu}</Box>
            </HStack>
            {bottomText(record)}
          </VStack>
        </ThemedBox>
        <PhotosLayout photos={photos} />
      </Pressable>
      <Divider />
    </>
  );
}
