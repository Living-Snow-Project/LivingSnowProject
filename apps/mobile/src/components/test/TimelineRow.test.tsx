import React from "react";
import { render } from "@testing-library/react-native";
import { makeExampleRecord } from "@livingsnow/record";
import { NativeBaseProviderForTesting } from "../../../jesttest.setup";
import { TimelineRow } from "../TimelineRow";
import { Labels } from "../../constants";

describe("TimelineRow test suite", () => {
  let expectedRecord;

  beforeEach(() => {
    expectedRecord = makeExampleRecord("Sample");
  });

  test("renders", () => {
    const { toJSON } = render(
      <NativeBaseProviderForTesting>
        <TimelineRow record={expectedRecord} />
      </NativeBaseProviderForTesting>
    );

    expect(toJSON()).toMatchSnapshot();
  });

  test("missing name", () => {
    expectedRecord.name = undefined;

    const { queryByText } = render(
      <NativeBaseProviderForTesting>
        <TimelineRow record={expectedRecord} />
      </NativeBaseProviderForTesting>
    );

    expect(queryByText(new RegExp(Labels.RecordFields.Name))).toBeFalsy();
  });

  test("missing organization", () => {
    expectedRecord.organization = undefined;

    const { queryByText } = render(
      <NativeBaseProviderForTesting>
        <TimelineRow record={expectedRecord} />
      </NativeBaseProviderForTesting>
    );

    expect(
      queryByText(new RegExp(Labels.RecordFields.Organization))
    ).toBeFalsy();
  });

  test("missing location description", () => {
    expectedRecord.locationDescription = undefined;

    const { queryByText } = render(
      <NativeBaseProviderForTesting>
        <TimelineRow record={expectedRecord} />
      </NativeBaseProviderForTesting>
    );

    expect(
      queryByText(new RegExp(Labels.RecordFields.LocationDescription))
    ).toBeFalsy();
  });

  test("missing notes", () => {
    expectedRecord.notes = undefined;

    const { queryByText } = render(
      <NativeBaseProviderForTesting>
        <TimelineRow record={expectedRecord} />
      </NativeBaseProviderForTesting>
    );

    expect(queryByText(new RegExp(Labels.RecordFields.Notes))).toBeFalsy();
  });
});
