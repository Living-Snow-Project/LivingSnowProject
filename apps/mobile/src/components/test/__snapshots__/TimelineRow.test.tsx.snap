// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimelineRow test suite renders 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    Array [
      Object {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    accessible={true}
    collapsable={false}
    dataSet={Object {}}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={Object {}}
    testID="1234"
  >
    <View
      dataSet={Object {}}
      style={
        Object {
          "paddingBottom": 4,
          "paddingLeft": 8,
          "paddingRight": 8,
          "paddingTop": 4,
        }
      }
    >
      <View
        dataSet={Object {}}
        style={Object {}}
      >
        <View
          dataSet={Object {}}
          style={
            Object {
              "flexDirection": "row",
            }
          }
        >
          <View
            dataSet={Object {}}
            style={
              Object {
                "width": "15%",
              }
            }
          >
            <View
              dataSet={Object {}}
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "#d8b4fe",
                  "borderColor": "#27272a",
                  "borderRadius": 9999,
                  "height": 48,
                  "justifyContent": "center",
                  "marginTop": 4,
                  "position": "relative",
                  "width": 48,
                }
              }
            >
              <Text
                dataSet={Object {}}
                style={
                  Object {
                    "backgroundColor": undefined,
                    "color": "#fafafa",
                    "fontFamily": undefined,
                    "fontSize": 16,
                    "fontStyle": "normal",
                    "fontWeight": "600",
                    "letterSpacing": 0,
                    "lineHeight": 24,
                    "textDecorationLine": undefined,
                  }
                }
              >
                TN
              </Text>
            </View>
          </View>
          <View
            dataSet={Object {}}
            style={
              Object {
                "paddingRight": 4,
                "width": "78%",
              }
            }
          >
            <View
              dataSet={Object {}}
              style={
                Object {
                  "marginLeft": 8,
                }
              }
            >
              <Text
                dataSet={Object {}}
                style={
                  Object {
                    "backgroundColor": undefined,
                    "color": "#171717",
                    "fontFamily": undefined,
                    "fontSize": 14,
                    "fontStyle": "normal",
                    "fontWeight": "600",
                    "letterSpacing": 0,
                    "lineHeight": 21,
                    "textDecorationLine": undefined,
                  }
                }
              >
                test name
              </Text>
              <View
                dataSet={Object {}}
                style={Object {}}
              />
              <Text
                dataSet={Object {}}
                style={
                  Object {
                    "backgroundColor": undefined,
                    "color": "#171717",
                    "fontFamily": undefined,
                    "fontSize": 14,
                    "fontStyle": "normal",
                    "fontWeight": "400",
                    "letterSpacing": 0,
                    "lineHeight": 21,
                    "textDecorationLine": undefined,
                  }
                }
              >
                test org
              </Text>
              <View
                dataSet={Object {}}
                style={Object {}}
              />
              <Text
                dataSet={Object {}}
                style={
                  Object {
                    "backgroundColor": undefined,
                    "color": "#db2777",
                    "fontFamily": undefined,
                    "fontSize": 14,
                    "fontStyle": "normal",
                    "fontWeight": "700",
                    "letterSpacing": 0,
                    "lineHeight": 21,
                    "textDecorationLine": undefined,
                  }
                }
              >
                Sample, 2021-09-16
              </Text>
            </View>
          </View>
          <View
            dataSet={Object {}}
            style={Object {}}
          />
          <View
            dataSet={Object {}}
            style={
              Object {
                "width": "7%",
              }
            }
          />
        </View>
        <View
          dataSet={Object {}}
          style={Object {}}
        />
        <Text
          dataSet={Object {}}
          style={
            Object {
              "backgroundColor": undefined,
              "color": "#171717",
              "fontFamily": undefined,
              "fontSize": 14,
              "fontStyle": "normal",
              "fontWeight": "500",
              "letterSpacing": 0,
              "lineHeight": 21,
              "textDecorationLine": undefined,
            }
          }
        >
          test location
        </Text>
        <View
          dataSet={Object {}}
          style={Object {}}
        />
        <Text
          dataSet={Object {}}
          style={
            Object {
              "backgroundColor": undefined,
              "color": "#171717",
              "fontFamily": undefined,
              "fontSize": 14,
              "fontStyle": "normal",
              "fontWeight": "400",
              "letterSpacing": 0,
              "lineHeight": 21,
              "textDecorationLine": undefined,
            }
          }
        >
          test notes
        </Text>
      </View>
    </View>
    <View
      dataSet={Object {}}
      style={
        Object {
          "flexDirection": "row",
        }
      }
    >
      <View
        dataSet={Object {}}
        style={
          Object {
            "alignItems": "center",
            "height": 2925,
            "justifyContent": "center",
            "width": 372,
          }
        }
      >
        <ActivityIndicator
          accessibilityLabel="loading"
          accessible={true}
          color="#0891b2"
          size="large"
          style={
            Array [
              Array [
                Object {},
                Object {
                  "dataSet": Object {},
                },
                Object {},
              ],
              undefined,
            ]
          }
        />
        <Text
          dataSet={Object {}}
          style={
            Object {
              "backgroundColor": undefined,
              "color": "#171717",
              "fontFamily": undefined,
              "fontSize": 14,
              "fontStyle": "normal",
              "fontWeight": "400",
              "letterSpacing": 0,
              "lineHeight": 21,
              "textDecorationLine": undefined,
            }
          }
        >
          Loading
        </Text>
      </View>
      <View
        dataSet={Object {}}
        style={Object {}}
      />
      <View
        dataSet={Object {}}
        style={
          Object {
            "paddingRight": 6,
          }
        }
      />
      <View
        dataSet={Object {}}
        style={Object {}}
      />
      <View
        dataSet={Object {}}
        style={
          Object {
            "alignItems": "center",
            "height": 2925,
            "justifyContent": "center",
            "width": 372,
          }
        }
      >
        <ActivityIndicator
          accessibilityLabel="loading"
          accessible={true}
          color="#0891b2"
          size="large"
          style={
            Array [
              Array [
                Object {},
                Object {
                  "dataSet": Object {},
                },
                Object {},
              ],
              undefined,
            ]
          }
        />
        <Text
          dataSet={Object {}}
          style={
            Object {
              "backgroundColor": undefined,
              "color": "#171717",
              "fontFamily": undefined,
              "fontSize": 14,
              "fontStyle": "normal",
              "fontWeight": "400",
              "letterSpacing": 0,
              "lineHeight": 21,
              "textDecorationLine": undefined,
            }
          }
        >
          Loading
        </Text>
      </View>
    </View>
  </View>
  <View
    dataSet={Object {}}
    style={
      Object {
        "backgroundColor": "#d6d3d1",
        "height": 12,
        "width": "100%",
      }
    }
  />
</RNCSafeAreaProvider>
`;
